File: packages/python/src/brainpick/querylog.py
"""The query log (spec/70): every brain_search call, raw, one JSON line per call, one file
per session — so a month later the question "do agents search with sentences or with
keyword lists?" is answered from the record, not from memory. Nothing is aggregated here.

Location: $BRAINPICK_QUERY_LOG_DIR, else $XDG_STATE_HOME/brainpick/queries (default
~/.local/state/brainpick/queries), file <session_id>.jsonl. BRAINPICK_QUERY_LOG=0 disables.
The session id is the MCP server process's (one per agent session under stdio) or the one a
CLI caller passes with --session, so a harness hook can log under its own session."""
from __future__ import annotations

import json
import os
import uuid
from datetime import datetime, timezone
from pathlib import Path


def new_session_id() -> str:
    return datetime.now(timezone.utc).strftime("%Y%m%dT%H%M%S") + "-" + uuid.uuid4().hex[:8]


def log_dir() -> Path | None:
    if os.environ.get("BRAINPICK_QUERY_LOG", "1") in ("0", "false", "off", ""):
        return None
    override = os.environ.get("BRAINPICK_QUERY_LOG_DIR")
    if override:
        return Path(override)
    state = os.environ.get("XDG_STATE_HOME") or str(Path.home() / ".local" / "state")
    return Path(state) / "brainpick" / "queries"


def log_query(session_id: str, brain: str, request: dict, result: dict) -> Path | None:
    """Append one line; never raises (a log must not break a search)."""
    directory = log_dir()
    if directory is None:
        return None
    line = {
        "ts": datetime.now(timezone.utc).isoformat(timespec="seconds"),
        "brain": brain,
        **request,
        "used_modes": result.get("used_modes"),
        "degraded_from": result.get("degraded_from"),
        "hits": [h.get("path") for h in result.get("hits", [])],
    }
    try:
        directory.mkdir(parents=True, exist_ok=True)
        path = directory / f"{session_id}.jsonl"
        with path.open("a", encoding="utf-8") as fh:
            fh.write(json.dumps(line, ensure_ascii=False) + "\n")
        return path
    except OSError:
        return None

File: packages/node/src/querylog.ts
/** The query log (spec/70): every brain_search call, raw, one JSON line per call, one file
 * per session — so a month later the question "do agents search with sentences or with
 * keyword lists?" is answered from the record, not from memory. Nothing is aggregated here.
 *
 * Location: $BRAINPICK_QUERY_LOG_DIR, else $XDG_STATE_HOME/brainpick/queries (default
 * ~/.local/state/brainpick/queries), file <session_id>.jsonl. BRAINPICK_QUERY_LOG=0 disables.
 * The session id is the MCP server process's (one per agent session under stdio) or the one
 * a CLI caller passes with --session, so a harness hook can log under its own session.
 * Mirrors brainpick/querylog.py. */
import { randomUUID } from "node:crypto";
import { appendFileSync, mkdirSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";

export function newSessionId(): string {
  const stamp = new Date().toISOString().replace(/[-:]/g, "").slice(0, 15); // YYYYMMDDTHHMMSS
  return `${stamp}-${randomUUID().replace(/-/g, "").slice(0, 8)}`;
}

export function logDir(): string | null {
  if (["0", "false", "off", ""].includes(process.env["BRAINPICK_QUERY_LOG"] ?? "1")) return null;
  const override = process.env["BRAINPICK_QUERY_LOG_DIR"];
  if (override) return override;
  const state = process.env["XDG_STATE_HOME"] || join(homedir(), ".local", "state");
  return join(state, "brainpick", "queries");
}

/** Append one line; never throws (a log must not break a search). */
export function logQuery(
  sessionId: string,
  brain: string,
  request: Record<string, unknown>,
  result: Record<string, unknown>,
): string | null {
  const directory = logDir();
  if (directory === null) return null;
  const hits = (result["hits"] as Array<Record<string, unknown>> | undefined) ?? [];
  const line = {
    ts: new Date().toISOString().replace(/\.\d{3}Z$/, "+00:00"),
    brain,
    ...request,
    used_modes: result["used_modes"] ?? null,
    degraded_from: result["degraded_from"] ?? null,
    hits: hits.map((h) => h["path"] ?? null),
  };
  try {
    mkdirSync(directory, { recursive: true });
    const path = join(directory, `${sessionId}.jsonl`);
    appendFileSync(path, JSON.stringify(line) + "\n", "utf8");
    return path;
  } catch {
    return null;
  }
}
